### ISSUE TITLE
Automatically rotate x-axis labels for ordinal x-axes

### ISSUE THREAD

**Issue by TuringLovesDeathMetal** [lightdash/lightdash#2028]
### Description: What is it?

For x-axes that are strings (ordinal values) we should rotate the labels 45 degrees once there are too many to display horizontally without overlapping.

Then, once the labels overlap at 45 degrees, we should rotate them to 90 degrees.

Once the 90 degrees values start overlapping, we should show no labels at all.

### Problem: What problem does this solve?

When there are too many x-axis values on a chart, then you end up with values that are missing. 
![image](https://user-images.githubusercontent.com/31848148/167386320-25588ce4-b25f-40e4-ad11-15c8faef7ee7.png)

This is fine if you're in Lightdash because you can hover over the values, but if you're sharing these charts externally, you can't see the values on the x-axis making the charts illegible (if they're ordinal values - i.e. not continuous ones like dates or continuous numbers)

### What: Roughly, what does this look like in the product?

- [ ] The chart has ordinal values on the x-axis (let's simplify this and just say `string` values)
- [ ] If the labels are overlapping being horizontal, we shift them to 45 degrees
- [ ] If the labels are overlapping at 45 degrees, we shift them to 90 degrees
- [ ] If the labels are overlapping at 90 degrees, we remove the labels

**Comment by TuringLovesDeathMetal** [comment]
related to #571 

Also, here's how they solved this problem at Metabase:
https://github.com/metabase/metabase/pull/7103
https://github.com/metabase/metabase/issues/7015

**Comment by TuringLovesDeathMetal** [comment]
We'll need to do some research into e-charts to see if this is feasible (can we detect when things are overlapping?)

**Comment by owlas** [comment]
@TuringLovesDeathMetal just to check, in the first example you have at the top of this issue, would you mind if the chart height was significantly reduced to make room for the 45 or 90 degree labels?

Or would you want to truncate the label lengths?

**Comment by TuringLovesDeathMetal** [comment]
@owlas I think we could shrink the chart a bit, but there would be a hard-limit on shrinking, then truncate the label lengths after that point. 

If this "hybrid" approach isn't possible, then let's just truncate the label lengths (not make the chart height way smaller)

**Comment by ZeRego** [comment]
Looks like metabase uses `http://dc-js.github.io/dc.js/` and has way more control over how the chart renders. They added a bunch of logic to calculate labels size and when to rotate https://github.com/metabase/metabase/pull/7103/files 
I can already tell this is going to be a challenge with echarts.

**Comment by PriPatel** [comment]
**Design feedback**

Instead of making the chart height smaller, can we make the chart panel bigger? I seem to remember a couple of people not being happy with how small the chart heights were in the first place!

**Comment by TuringLovesDeathMetal** [comment]
@PriPatel yep, that could be cool - but with some hard limit on the size. (other options could be that it stretches out if you close the other tabs, or there's a button you can click to see the chart `fullscreen`)

@ZeRego ya, that's what I was afraid of...it feels like we need to do some research into this. To start off, I'd love to know if in e-charts can we:

1. calculate label sizes
2. calculate pixel-width of bars
3. detect overlap in text
4. enable conditional rotation of labels (i.e. "rotate label if _x_ is TRUE")

**Comment by TuringLovesDeathMetal** [comment]
I wonder if there's some way that we can make the chart scroll-able? 🤔 This feels like a really non-ideal solution though...

https://github.com/lightdash/lightdash/issues/1623

**Comment by JohnRomanski** [comment]
Suggest that you make this an option that is configurable by the user: Horizontal, angled 45 degrees right, left, vertical.
Note that not all columns need labels, especially for time series.

**Comment by TuringLovesDeathMetal** [comment]
@JohnRomanski - ya, I think there should be a great default, but it should also be configurable. 

By default, for _**ordinal**_ axes (i.e. string axes labels) we would force all labels to show (up to some point where labels at 90 degrees would overlap). 

But, for _**any**_ axis type, you'd still have the option to configure the axis. <- this feels like a separate bit of work to the "sensible defaults"

**Comment by TuringLovesDeathMetal** [comment]
Opened a new issue for the manual configuration: https://github.com/lightdash/lightdash/issues/2272

**Comment by ZeRego** [comment]
There is a [echarts ticket for this feature request](https://github.com/apache/echarts/issues/14996)

**Comment by IrakliJani** [comment]
still relevant

**Comment by ZeRego** [comment]
What if we allowed the editor to manually set the rotation angle or hide all labels for each chart? 
This would unblock some of our users that know for sure their chart is going to have such problems. 
Just had feedback from a user that would rather hide all labels than have just a few visible.

**Comment by magnew** [comment]
A number of comments here have suggested we make this a configuration instead of automatic. Which we have done! The commit is here: https://github.com/lightdash/lightdash/pull/8657

I'm going to close this ticket as we have handled this scenario. We can think about whether we need to automatic version.
